### PR TITLE
[SPARK-20910][SQL] Add build-in SQL function - UUID

### PR DESCRIPTION
**Pull request by wangyum**
## What changes were proposed in this pull request?

Add build-int SQL function - UUID.

## How was this patch tested?

unit tests

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala [hunk 1]
```diff
@@ -104,3 +107,25 @@ case class CurrentDatabase() extends LeafExpression with Unevaluable {
...
+// scalastyle:off line.size.limit
+@ExpressionDescription(
+  usage = "_FUNC_() - Returns a universally unique identifier (UUID) string. The value is returned as a canonical UUID 36-character string.",
+  extended = """
+    Examples:
+      > SELECT _FUNC_();
+       46707d92-02f4-4817-8116-a4c3b23e6266
+  """)
+// scalastyle:on line.size.limit
+case class Uuid() extends LeafExpression with CodegenFallback {
```
Comment (ueshin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala]: Can you implement codegen for this?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala [hunk 2]
```diff
@@ -104,3 +107,25 @@ case class CurrentDatabase() extends LeafExpression with Unevaluable {
...
+  override def foldable: Boolean = true
```
Comment (ueshin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala]: I don't think this function is foldable. If this is foldable, it means we might use the same UUID generated by optimizer for all rows.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala [hunk 3]
```diff
@@ -104,3 +107,26 @@ case class CurrentDatabase() extends LeafExpression with Unevaluable {
...
+case class Uuid() extends LeafExpression {
+
+  override def nullable: Boolean = false
+
+  override def dataType: DataType = StringType
+
+  override def eval(input: InternalRow): Any = UTF8String.fromString(UUID.randomUUID().toString)
+
+  override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
+    ev.copy(code = s"final ${ctx.javaType(dataType)} ${ev.value} = " +
```
Comment (kiszk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala]: nit: ${ctx.javaType(dataType)} -> UTF8String

File: sql/core/src/test/resources/sql-tests/inputs/string-functions.sql [hunk 4]
```diff
@@ -14,4 +14,4 @@ select replace('abc', 'b', '123');
 select replace('abc', 'b');
 
 -- uuid
-select length(uuid()), (uuid() <> uuid());
```
Comment (ueshin) [inline sql/core/src/test/resources/sql-tests/inputs/string-functions.sql]: This test should work if `Uuid` is `deterministic = false`.

Comment (viirya) [inline sql/core/src/test/resources/sql-tests/inputs/string-functions.sql]: I think what @ueshin means is we should set `deterministic = false` for `Uuid` expression.

Comment (ueshin) [inline sql/core/src/test/resources/sql-tests/inputs/string-functions.sql]: Ah, yes, thank you for supporting.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala [hunk 5]
```diff
@@ -104,3 +107,28 @@ case class CurrentDatabase() extends LeafExpression with Unevaluable {
   override def nullable: Boolean = false
   override def prettyName: String = "current_database"
 }
+
+// scalastyle:off line.size.limit
+@ExpressionDescription(
+  usage = "_FUNC_() - Returns a universally unique identifier (UUID) string. The value is returned as a canonical UUID 36-character string.",
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala]: nit: ... an universally ...

**Comment by ueshin** [comment]
LGTM, pending tests.

**Comment by viirya** [comment]
LGTM

**Comment by ueshin** [comment]
Thanks! Merging to master.

**Comment by hvanhovell** [comment]
I just came across this expression and I have a few concerns:

1. A row will not get the same UUID assigned when a task fails. This might cause some really weird problems when the UUID column is used later-on.
2. UUID.randomString() uses a lot of synchronization in the background. This might make it pretty slow.

Shall I file a ticket?

**Pull request closed without merging**